lib/hostip.c

```diff
@@ -819,48 +819,95 @@ CURLcode Curl_loadhostpairs(struct Curl_easy *data)
     }
     else {
       struct Curl_dns_entry *dns;
-      Curl_addrinfo *addr;
+      Curl_addrinfo *head = 0, *tail = 0;
       char *entry_id;
       size_t entry_len;
-      char buffer[256];
-      char *address = &buffer[0];
-
-      if(3 != sscanf(hostp->data, "%255[^:]:%d:%255s", hostname, &port,
-                     address)) {
-        infof(data, "Couldn't parse CURLOPT_RESOLVE entry '%s'!\n",
-              hostp->data);
-        continue;
+      char address[64];
+      char *addrPointer;
+      char *last;
+      char *tmp;
+      unsigned long tmpPort;
+      int error = true;
+
+      tmp = strchr(hostp->data, ':');
+      if(!tmp || (tmp - hostp->data) >= (ptrdiff_t)sizeof(hostname)) {
+        goto err;
       }
+      memcpy(hostname, hostp->data, tmp - hostp->data);
+      hostname[tmp - hostp->data] = '\0';
 
-      /* allow IP(v6) address within [brackets] */
-      if(address[0] == '[') {
-        size_t alen = strlen(address);
-        if(address[alen-1] != ']')
-          /* it needs to also end with ] to be valid */
-          continue;
-        address[alen-1] = 0; /* zero terminate there */
-        address++; /* pass the open bracket */
+      tmpPort = strtoul(tmp + 1, &tmp, 10);
+      if(!tmpPort || tmpPort > USHRT_MAX || *tmp != ':' || !(tmp + 1)) {
+        goto err;
       }
+      port = (int)tmpPort;
+
+      while(true) {
+        size_t alen;
+
+        last = tmp + 1;
+        tmp = strchr(last, ',');
+        if(tmp)
+          alen = tmp - last;
+        else
+          alen = strlen(hostp->data) - (last - hostp->data);
```

Comment:
i don't follow this
~~~
foo.bar:80:1.1.1.1,2.2.2.2,3.3.3.3
           ^      ^
           |      |
           -last  -tmp
~~~
assume `hostp->data` is 0 just for pseudo math ehre
34 - (11 - 0) = 23 address len?

also personally i find last and tmp confusing, they could be begin and end.

Reply:
I just needed to get the pointer converted into an offset but I now realize it's better to just do strlen addrBegin

I also renamed all the variables to more readable one.

Reply:
I just needed to do the pointer arithmetic to get the offset but I changed it to strlen(addrEnd) which reads better and is faster anyway.